Permute receivers (uniformly) randomly

akalin-keybase opened the pull request:
This prevents recipients from being able
to infer data about other recipients based
on order.

akalin-keybase commented on signcrypt_seal_test.go:
```diff
@@ -0,0 +1,35 @@
+)
+
+func TestSealEncryptionKeyForReceiversPermuted(t *testing.T) {
+	const count = 10
```
TODO: Flesh this test out a bit more

akalin-keybase commented on signcrypt_seal.go:
```diff
@@ -246,6 +236,24 @@ func (sss *signcryptSealStream) init(receiverBoxKeys []BoxPublicKey, receiverSym
+	// all of them.
+	for i, receiverBoxKey := range receiverBoxKeys {
+		index := order[i]
+		receiverKeysArray[index] = receiverEntryForBoxKey(receiverBoxKey, ephemeralKey, encryptionKey, uint64(index))
```
Ah, the identifier depends on the index, unlike the encryption case. Does this imply quadratic behavior on the decryption side, like you mentioned?

oconnor663 replied:
I think so, though currently the number of receiver box secret keys you're going to have is one.

oconnor663 replied:
And even when it's more than one (shared DH keys?), the number of recipients on a message is also expected to be "small".

akalin-keybase commented:
Looks like the coverage decrease is due to some added panics for switch cases.

akalin-keybase commented on encrypt.go:
```diff
@@ -187,17 +175,64 @@ func (es *encryptStream) init(version Version, sender BoxSecretKey, receivers []
+		}
+		return order
+	case 2:
+		return randomPerm(receiverCount)
```
Oh duh does it even make sense to permute the order for encryption mode? The IDs are already public, right?

oconnor663 replied:
By default yes, though there's an anonymous mode that could benefit.

oconnor663 commented on common.go:
```diff
@@ -6,10 +6,11 @@ package saltpack
 	"bytes"
 	"crypto/hmac"
-	"crypto/rand"
+	cryptorand "crypto/rand"
```
This is a good idea. I'll start doing it elsewhere.

akalin-keybase replied:
I did it here out of necessity (since I use both) but yeah maybe it's a good idea in general.

oconnor663 commented on common.go:
```diff
@@ -37,6 +38,27 @@ func randomFill(b []byte) (err error) {
+	var buf [8]byte
+	cryptorand.Read(buf[:])
+	return int64(binary.BigEndian.Uint64(buf[:]) >> 1)
+}
```
Curious what happens if this function actually returns an int > 2^63...

akalin-keybase replied:
You mean a negative number? I think callers in `math.rand` would panic.

oconnor663 commented on signcrypt_seal_test.go:
```diff
@@ -0,0 +1,35 @@
+	receiverKeysArray2 := sealEncryptionKeyForReceivers(receiverBoxKeys, receiverSymmetricKeys, ephemeralKey, encryptionKey)
+
+	// Technically this check is flaky, but the flake probability
+	// is 1/20! ~ 2^{-61}.
```
lol

oconnor663 commented on encrypt.go:
```diff
@@ -189,15 +176,57 @@ func (es *encryptStream) init(version Version, sender BoxSecretKey, receivers []
 
+func computeReceiverOrder(version Version, receiverCount int) []int {
+	switch version.Major {
+	case 1:
```
I don't think randomizing the receivers will break anything in v1. Any reason not to do it for both?

akalin-keybase replied:
Yeah, you're right. Done.

oconnor663 commented on signcrypt_seal.go:
```diff
@@ -250,6 +240,24 @@ func (sss *signcryptSealStream) init(receiverBoxKeys []BoxPublicKey, receiverSym
+	for i, receiverSymmetricKey := range receiverSymmetricKeys {
+		index := order[len(receiverBoxKeys)+i]
+		receiverKeysArray[index] = receiverEntryForSymmetricKey(receiverSymmetricKey, ephemeralKey.GetPublicKey(), encryptionKey, uint64(index))
+	}
```
Would it be easier to try to define a function that permuted an array in place? Even if we had to copy it for different data types?

akalin-keybase replied:
Yeah, that turns out to be much simpler! I implemented that.

akalin-keybase commented:
Addressed comments, which pretty much caused a rewrite, although it's much simpler now!

I think it's worth you taking another quick review. Thanks!

oconnor663 commented on signcrypt_seal.go:
```diff
@@ -180,16 +190,29 @@ func receiverEntryForSymmetricKey(receiverSymmetricKey ReceiverSymmetricKey, eph
 
+func shuffleSigncryptionReceivers(receiverBoxKeys []BoxPublicKey, receiverSymmetricKeys []ReceiverSymmetricKey) []receiverKeysMaker {
+	order := randomPerm(len(receiverBoxKeys) + len(receiverSymmetricKeys))
+	receivers := make([]receiverKeysMaker, len(receiverBoxKeys)+len(receiverSymmetricKeys))
```
could compute `totalLen` once maybe, to make it easier to read?
